waku/v2/api/publish/message_sender.go

```diff
@@ -0,0 +1,124 @@
+package publish
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/waku-org/go-waku/waku/v2/protocol"
+	"github.com/waku-org/go-waku/waku/v2/protocol/lightpush"
+	"github.com/waku-org/go-waku/waku/v2/protocol/relay"
+	"go.uber.org/zap"
+	"golang.org/x/time/rate"
+)
+
+const DefaultPeersToPublishForLightpush = 2
+const DefaultPublishingLimiterRate = rate.Limit(2)
+const DefaultPublishingLimitBurst = 4
+
+type PublishMethod int
+
+const (
+	LightPush PublishMethod = iota
+	Relay
+)
+
+func (pm PublishMethod) String() string {
+	switch pm {
+	case LightPush:
+		return "LightPush"
+	case Relay:
+		return "Relay"
+	default:
+		return "Unknown"
+	}
+}
+
+type MessageSender struct {
+	ctx              context.Context
+	publishMethod    PublishMethod
+	lightPush        *lightpush.WakuLightPush
+	relay            *relay.WakuRelay
+	messageSentCheck *MessageSentCheck
+	rateLimiter      *PublishRateLimiter
+	logger           *zap.Logger
+}
+
+func NewMessageSender(ctx context.Context, publishMethod PublishMethod, lightPush *lightpush.WakuLightPush, relay *relay.WakuRelay, logger *zap.Logger) *MessageSender {
+	return &MessageSender{
+		ctx:           ctx,
+		publishMethod: publishMethod,
+		lightPush:     lightPush,
+		relay:         relay,
+		rateLimiter:   NewPublishRateLimiter(DefaultPublishingLimiterRate, DefaultPublishingLimitBurst),
+		logger:        logger,
+	}
+}
+
+func (ms *MessageSender) WithMessageSentCheck(messageSentCheck *MessageSentCheck) *MessageSender {
+	ms.messageSentCheck = messageSentCheck
+	return ms
+}
+
+func (ms *MessageSender) WithRateLimiting(rateLimiter *PublishRateLimiter) *MessageSender {
+	ms.rateLimiter = rateLimiter
+	return ms
+}
+
+func (ms *MessageSender) Send(env *protocol.Envelope) error {
```

Comment:
What do you think of instead using the context from the struct, to use instead a context passed as a parameter in this function? that way you can make the Send function cancellable.

Reply:
Thanks for the tips!